File: leetcode/algorithms/102_binary_tree_level_order_traversal/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * BFS
 *
 * Complexities:
 *   N - Size of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(N)
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  const result = [];

  if (!root) {
    return result;
  }

  const q = [];
  q.push(root);

  while (q.length > 0) {
    const size = q.length;
    const array = [];

    for (let i = 0; i < size; i++) {
      const node = q.shift();

      array.push(node.val);

      if (node.left) {
        q.push(node.left);
      }
      if (node.right) {
        q.push(node.right);
      }
    }

    result.push(array);
  }

  return result;
};


// Solution
/**
 * Solution 1
 *
 * BFS
 *
 * Complexities:
 *   N - Size of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(N)
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var solution1 = function (root) {
  if (!root) {
    return [];
  }

  const result = [];
  const queue = [root];

  while (queue.length > 0) {
    const levelSize = queue.length;
    const currentLevel = [];

    for (let i = 0; i < levelSize; i++) {
      const currentNode = queue.shift();
      currentLevel.push(currentNode.val);

      if (currentNode.left) {
        queue.push(currentNode.left);
      }
      if (currentNode.right) {
        queue.push(currentNode.right);
      }
    }

    result.push(currentLevel);
  }

  return result;
};

/**
 * Solution 2
 *
 * DFS
 *
 * Complexities:
 *   N - Size of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(N)
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var solution2 = function (root) {
  const result = [];

  function dfs(node, level) {
    if (!node) {
      return;
    }

    if (result.length === level) {
      result.push([]);
    }

    result[level].push(node.val);

    dfs(node.left, level + 1);
    dfs(node.right, level + 1);
  }

  dfs(root, 0);
  return result;
};

File: leetcode/algorithms/102_binary_tree_level_order_traversal/main.py
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTreeLevelOrderTraversal:
    """
    # BFS
    #
    # Complexities:
    #   N - Size of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result: List[List[int]] = list()

        if root == None:
            return result

        q: deque = deque()
        q.append(root)

        while len(q) > 0:
            array: List[int] = list()

            for _ in range(0, len(q)):
                node: Optional[TreeNode] = q.popleft()

                array.append(node.val)

                if node.left != None:
                    q.append(node.left)
                if node.right != None:
                    q.append(node.right)

            result.append(array)

        return result


    # Solution
    """
    # Solution 1
    #
    # BFS
    #
    # Complexities:
    #   N - Size of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def solution1(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        result = []
        queue = deque([root])

        while queue:
            level_size = len(queue)
            current_level = []

            for _ in range(level_size):
                node = queue.popleft()
                current_level.append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(current_level)

        return result

    """
    # Solution 2
    #
    # DFS
    #
    # Complexities:
    #   N - Size of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def solution2(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []

        def dfs(node, level):
            if not node:
                return

            if len(result) == level:
                result.append([])

            result[level].append(node.val)

            dfs(node.left, level + 1)
            dfs(node.right, level + 1)

        dfs(root, 0)
        return result
